Firebend.AutoCrud.Core/Implementations/Entities/ClientRequestTransactionManager.cs

```diff
@@ -0,0 +1,84 @@
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Firebend.AutoCrud.Core.Interfaces.Models;
+using Firebend.AutoCrud.Core.Interfaces.Services.Entities;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Firebend.AutoCrud.Core.Implementations.Entities;
+
+public class ClientRequestTransactionManager : ISessionTransactionManager
+{
+    private readonly IServiceProvider _serviceProvider;
+
+    private readonly ConcurrentDictionary<Type, Task<IEntityTransaction>> _sharedTransactions = new();
+    private readonly ConcurrentBag<IEntityTransaction> _transactions = new();
+    public bool TransactionStarted { get; private set; }
+    public ImmutableList<Guid> TransactionIds => _transactions.Select(x => x.Id).ToImmutableList();
+
+    public ClientRequestTransactionManager(IServiceProvider serviceProvider)
+    {
+        _serviceProvider = serviceProvider;
+    }
+
+    public void Start() => TransactionStarted = true;
+
+    public async Task CompleteAsync(CancellationToken cancellationToken)
+    {
+        await Task.WhenAll(_transactions.Select(x => x.CompleteAsync(cancellationToken)));
+        ClearTransactions();
+    }
+
+    public async Task RollbackAsync(CancellationToken cancellationToken)
+    {
+        await Task.WhenAll(_transactions.Select(x => x.RollbackAsync(cancellationToken)));
+        ClearTransactions();
+    }
+
+    public async Task<IEntityTransaction> GetTransaction<TKey, TEntity>(CancellationToken cancellationToken)
+        where TKey : struct
+        where TEntity : IEntity<TKey>
+    {
+        if (!TransactionStarted)
+        {
+            return null;
+        }
+
+        var transactionFactory = _serviceProvider.GetRequiredService<IEntityTransactionFactory<TKey, TEntity>>();
+        return await _sharedTransactions.GetOrAdd(transactionFactory.GetType(),
+            async (_) =>
+            {
+                var transaction = await transactionFactory.StartTransactionAsync(cancellationToken);
+                _transactions.Add(transaction);
+                return transaction;
+            });
+    }
+
+    public void AddTransaction(IEntityTransaction transaction)
```

Comment:
for convenience, should we add a `AddTransaction<TKey, TEntity>` that calls GetTransaction and then passes that to AddTransaction?

Reply:
I'm not sure I understand since GetTransaction currently will add the transaction

Reply:
oh i didn't notice that. then do we need to pass it into .addtransaction in the services? could we just pass null?

Reply:
i could still see that method being useful if i just want to add one and don't care about getting it back in the return value though